File: src/C/scale_matrix.c
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-16 19:12:24
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-16 19:15:26
@FilePath: \asm_matrix_benchmark\src\C\scale_matrix.c
@Description: scale matrix c code
*************************************************************/
#include "scale_matrix.h"

/***********************************************************
@description: scale matrix
@param {MatrixInt} *m
@param {int} scale
@return {*}
 ************************************************************/
MatrixInt *scale_matrix_int(MatrixInt *m, int scale)
{
    // check m
    if (m == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // check m->data
    if (m->data == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // malloc for new res
    MatrixInt *res = NULL;
    res = (MatrixInt *)malloc(sizeof(MatrixInt));
    if (res == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    res->cols = m->cols;
    res->rows = m->rows;
    // malloc for res->data
    res->data = (int *)malloc(sizeof(int) * res->cols * res->rows);
    if (res->data == NULL)
    {
        free(res);
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    for (size_t i = 0; i < res->cols * res->rows; i++)
    {
        res->data[i] = m->data[i] * scale;
    }
    return res;
}

/***********************************************************
@description: scale matrix float
@param {MatrixFloat} *m
@param {float} scale
@return {*}
************************************************************/
MatrixFloat *scale_matrix_float(MatrixFloat *m, float scale)
{
    // check m
    if (m == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // check m->data
    if (m->data == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // malloc for new res
    MatrixFloat *res = NULL;
    res = (MatrixFloat *)malloc(sizeof(MatrixFloat));
    if (res == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    res->cols = m->cols;
    res->rows = m->rows;
    // malloc for res->data
    res->data = (float *)malloc(sizeof(float) * res->cols * res->rows);
    if (res->data == NULL)
    {
        free(res);
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    for (size_t i = 0; i < res->cols * res->rows; i++)
    {
        res->data[i] = m->data[i] * scale;
    }
    return res;
}

File: src/JavaScript/scale_matrix.js
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-29 20:10:32
@LastEditors: GitHubCopilot
@LastEditTime: 2026-06-29 20:10:32
@FilePath: \asm_matrix_benchmark\src\JavaScript\scale_matrix.js
@Description: scale matrix JavaScript code
*************************************************************/

"use strict";

const { checkMatrix, buildTypedMatrix } = require("./common");

/**
 * Multiply all matrix elements by scalar.
 * @param {{rows:number, cols:number, data:number[]}} m
 * @param {number} scalar
 * @param {"int"|"float"|"double"} type
 * @returns {{rows:number, cols:number, data:number[]}|null}
 */
function scaleMatrix(m, scalar, type) {
  if (
    !checkMatrix(m, "m") ||
    typeof scalar !== "number" ||
    Number.isNaN(scalar)
  ) {
    console.error("Invalid param!");
    return null;
  }

  const data = new Array(m.rows * m.cols);
  for (let i = 0; i < data.length; i++) {
    data[i] = m.data[i] * scalar;
  }

  return buildTypedMatrix(m.rows, m.cols, data, type);
}

function scale_matrix_int(m, scalar) {
  return scaleMatrix(m, scalar, "int");
}

function scale_matrix_float(m, scalar) {
  return scaleMatrix(m, scalar, "float");
}

function scale_matrix_double(m, scalar) {
  return scaleMatrix(m, scalar, "double");
}

module.exports = {
  scale_matrix_int,
  scale_matrix_float,
  scale_matrix_double,
};
